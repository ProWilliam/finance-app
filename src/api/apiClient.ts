// Config and Types
import config from '../../app.config';
import { FetchOptions } from '../types/api/Api.types';

const API_URL = config.extra.apiUrl;

const apiClient = async (endpoint: string, options?: FetchOptions) => {
  const { method = 'GET', headers = {}, body } = options || {};

  const config: RequestInit = {
    method,
    headers: {
      'Content-Type': 'application/json',
      ...headers,
    },
    body: JSON.stringify(body),
  };

  return globalThis.fetch(`${API_URL}${endpoint}`, config)
    .then(response => {
      if (!response.ok) {
        throw new Error(`Error: ${response.statusText}`);
      }
      return response.json();
    })
    .catch(error => {
      throw error;
    });

};

export default apiClient;