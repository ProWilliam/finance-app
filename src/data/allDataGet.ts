export default {
    id: '1', 
    infoText: 'Informacion extra', 
    name: 'Cart black', 
    description: 'Descripcion registrada del producto', 
    logo: 'https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg', 
    dataRelease: new Date('2025-03-02'), 
    dataRevision: new Date('2026-08-15')
  }
