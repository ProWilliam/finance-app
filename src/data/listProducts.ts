export default [
  { id: 1, name: 'Nombre1' },
  { id: 2, name: 'Nombre2' },
  { id: 3, name: 'Nombre3' },
  { id: 4, name: 'Nombre4' },
  { id: 5, name: 'Nombre5' },
  { id: 6, name: 'Nombre6' },
  { id: 8, name: 'Nombre8' },
  { id: 9, name: 'Nombre9' },
  { id: 10, name: 'Nombre10' },
  { id: 11, name: 'Nombre11' },
  { id: 12, name: 'Nombre12' },
  { id: 13, name: 'Nombre13' },
  { id: 14, name: 'Nombre14' },
  { id: 15, name: 'Nombre13' },
  { id: 16, name: 'Nombre13' },
  { id: 17, name: 'Nombre13' },
  { id: 18, name: 'Nombre13' },
  { id: 19, name: 'Nombre13' },
  { id: 20, name: 'Nombre13' },
]