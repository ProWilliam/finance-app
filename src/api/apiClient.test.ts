import apiClient from './apiClient';
import config from '../../app.config';
import data from '../data/allDataGet';

// Mock by function global fetch
globalThis.fetch = jest.fn() as jest.MockedFunction<typeof fetch>;

describe('apiclient', () => {

  
  let mockData: {data: typeof data[]};
  let routeApi: string;
  let allRouteApi: string;
  let id: string;
  let postData: {data: typeof data}

  // clean data before each test.
  beforeEach(() => {
    (globalThis.fetch as jest.Mock).mockClear();
    jest.useFakeTimers();

    mockData = { data: [data]}; // Import format per server
    postData = { data: data}; 

    routeApi = config.extra.productUrl;
    allRouteApi = config.extra.apiUrl + config.extra.productUrl;
    id = '1'
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllTimers();
    jest.restoreAllMocks();
  })

  it('Should handle errors gracefully', async () => {
    (globalThis.fetch as jest.Mock).mockRejectedValueOnce(new Error('Failed to fetch'));

    await expect(apiClient('/test-endpoint')).rejects.toThrow('Failed to fetch');
  });

  it('Should perform a Get request and return data', async () => {
    

    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({

      ok: true,
      json: async () => mockData,

    } as Response );

    const response = await apiClient(routeApi);

    expect(response).toEqual(mockData);
    expect(globalThis.fetch).toHaveBeenCalledWith(allRouteApi, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    });

    expect(response).toMatchSnapshot();

  });

  it('Should perform a POST request with body', async () => {

    const mockData = { success: true };
    
    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => mockData,
    });

    
    const response = await apiClient(routeApi, {
      method: 'POST',
      body: postData,
    });

    expect(response).toEqual(mockData);
    expect(globalThis.fetch).toHaveBeenCalledWith(
      allRouteApi,
      expect.objectContaining({
        method: 'POST',
        body: JSON.stringify(postData),
      })
    );

    expect(response).toMatchSnapshot();

  });

  it('Should perform a PUT request with body', async () => {

    const mockData = {
      id: "1",
      name: "Updated Product",
      description: "Updated Description",
      logo: "updated-logo.png",
      date_release: "2025-01-01",
      date_revision: "2025-01-01"
    };

    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => mockData,
    });

    
    const response = await apiClient(routeApi + id, {
      method: 'PUT',
      body: mockData 
    });

    expect(response).toEqual(mockData);
    expect(globalThis.fetch).toHaveBeenCalledWith(
      allRouteApi + id,
      expect.objectContaining({
        method: 'PUT',
        body: JSON.stringify(mockData),
      })
    );

    expect(response).toMatchSnapshot();

  });

  it('Should send a DELETE method with id', async () => {

    const mockData = { success: true };

    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => mockData,
    });

    
    const response = await apiClient(routeApi + id, {
      method: 'DELETE',
    });

    expect(response).toEqual(mockData);
    expect(globalThis.fetch).toHaveBeenCalledWith(
      allRouteApi + id,
      expect.objectContaining({
        method: 'DELETE',
      })
    );

    expect(response).toMatchSnapshot();
  });

  it('Should handle invalid data in response', async () => {
    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => ({ unexpectedKey: 'unexpectedValue' }),
    });
  
    const response = await apiClient(routeApi);
    expect(response).not.toEqual(expect.objectContaining({ data: expect.anything() }));
  });

  it('Should send custom headers correctly', async () => {
    const customHeaders = { Authorization: 'TokenHASH256' };
  
    (globalThis.fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => ({}),
    });
  
    await apiClient(routeApi, {
      method: 'GET',
      headers: customHeaders,
    });
  
    expect(globalThis.fetch).toHaveBeenCalledWith(
      expect.anything(),
      expect.objectContaining({
        headers: expect.objectContaining(customHeaders),
      })
    );
  });

  it('Should handle multiple requests concurrently', async () => {
    (globalThis.fetch as jest.Mock)
      .mockResolvedValueOnce({ ok: true, json: async () => ({ id: 1 }) })
      .mockResolvedValueOnce({ ok: true, json: async () => ({ id: 2 }) });
  
    const request1 = apiClient(routeApi);
    const request2 = apiClient(routeApi);
  
    const responses = await Promise.all([request1, request2]);
  
    expect(responses).toEqual([{ id: 1 }, { id: 2 }]);
    expect(globalThis.fetch).toHaveBeenCalledTimes(2);
  });

  it('Should handle timeouts correctly', async () => {
    
    (globalThis.fetch as jest.Mock).mockImplementationOnce(() =>
      new Promise((resolve) => globalThis.setTimeout(() => resolve({ ok: true, json: async () => ({}) }), 5000))
    );
    

    const fetchPromise = apiClient(routeApi);

    jest.advanceTimersByTime(2500);

    let isPending = true;
    fetchPromise.then(() => {
      isPending = false;
    });
  
    expect(isPending).toBe(true);
  
    jest.advanceTimersByTime(2500);
  
    await expect(fetchPromise).resolves.toBeDefined();
    expect(isPending).toBe(false);
  });

});