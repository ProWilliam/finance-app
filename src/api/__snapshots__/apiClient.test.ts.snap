// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiclient Should perform a Get request and return data 1`] = `
{
  "data": [
    {
      "date_release": "2025-03-02",
      "date_revision": "2026-08-15",
      "description": "Descripcion registrada del producto",
      "id": "1",
      "info_text": "Informacion extra",
      "logo": "https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg",
      "name": "Cart black",
    },
  ],
}
`;

exports[`apiclient Should perform a POST request with body 1`] = `
{
  "success": true,
}
`;

exports[`apiclient Should perform a PUT request with body 1`] = `
{
  "date_release": "2025-01-01",
  "date_revision": "2025-01-01",
  "description": "Updated Description",
  "id": "1",
  "logo": "updated-logo.png",
  "name": "Updated Product",
}
`;

exports[`apiclient Should send a DELETE method with id 1`] = `
{
  "success": true,
}
`;
